### PR TITLE
Only display JavaScript inlay hints for code in view

### LINKED ISSUE #3459: Inlay hints added to wrong buffer when original buffer is not active (lsp-javascript, lsp-rust)
### Thank you for the bug report

- [X] I am using the latest version of `lsp-mode` related packages.
- [X] I checked [FAQ](https://emacs-lsp.github.io/lsp-mode/page/faq/) and [Troubleshooting](https://emacs-lsp.github.io/lsp-mode/page/troubleshooting/) sections
- [X] You may also try reproduce the issue using clean environment using the following command: `M-x lsp-start-plain`

### Bug description

#3455 inherited a bug from #3404 where it’s possible for inlay hints to be added to the wrong buffer. For example, watch the Magit buffer in this recording:

Inlay hints are incorrectly added to the active (Magit) buffer instead of the relevant buffer

This is happening because the handler for the async request is operating on whatever the active buffer is.

### Steps to reproduce

These two steps must be performed in quick succession:

1. Open a JavaScript or Rust file while having `lsp-javascript-display-hints` or `lsp-rust-analyzer-server-display-hints` (respectively) set to `t`.
2. Immediately switch to another buffer.

### Expected behavior

Inlay hints are added to the buffer they belong to.

### Which Language Server did you use?

lsp-javascript
lsp-rust

### OS

Windows

### Error callstack

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by shivjm**
This drastically speeds up working with large files. A work project has a 90,000-line JavaScript file. Before, I couldn’t edit it with inlay hints enabled since it would just keep locking up (even after the initial delay while it was being parsed and everything). In contrast, with this change, inlay hints have virtually no impact on performance regardless of the size of the file. In fact, I tried making the hints more aggressive by updating them when scrolling or changing the size of the window and noticed no impact on responsiveness. (I added a custom variable to allow disabling this in case it does impact performance on other machines.)

Bonus:
Fixes #3459

### CODE REVIEW

File: clients/lsp-javascript.el [hunk 1]
```diff
@@ -817,28 +822,49 @@ name (e.g. `data' variable passed as `data' parameter)."
...
-                       (overlay (make-overlay pos pos nil 'front-advance 'end-advance)))
-                 (overlay-put overlay 'lsp-javascript-inlay-hint t)
-                 (overlay-put overlay 'before-string
-                              (format "%s%s%s"
-                                      (if (and whitespace-before? (not (string= kind lsp/javascript-inlay-hint-kind-type-hint))) " " "")
-                                      (propertize (lsp-javascript-format-inlay text kind)
-                                                  'font-lock-face (lsp-javascript-face-for-inlay kind))
-                                      (if whitespace-after? " " ""))))))))
+         (let ((hints (lsp-get res :inlayHints)))
+           (when (> (length hints) 0)
```
Comment (shivjm) [inline clients/lsp-javascript.el:844]: Open to suggestions for how to phrase this entire `when` block. The idea is to call `overlay-recenter` only if there are hints to add.

File: lsp-protocol.el [hunk 2]
```diff
@@ -425,7 +425,8 @@ See `-let' for a description of the destructuring mechanism."
 (defconst lsp/javascript-inlay-hint-kind-parameter-hint "Parameter")
 (defconst lsp/javascript-inlay-hint-kind-enum-hint "Enum")
 (lsp-interface (javascript:InlayHint (:text :position :kind) (:whitespaceBefore :whitespaceAfter))
-               (javascript:InlayHintsParams (:textDocument) (:range)))
+               (javascript:InlayHintsParams (:textDocument) (:range))
+               (javascript:Range (:start :end)))
```
Comment (yyoncho) [inline lsp-protocol.el:429]: there is no need to define that range, the normal range should work fine.

**Comment by shivjm** [comment]
I removed the `window-size-change-functions` function because I kept seeing errors like this in `*Messages*`:
```
Error during redisplay: (lsp-javascript-update-inlay-hints-size-change-function #<window 3 on file.js>) signaled (wrong-type-argument stringp nil)
```
It didn’t seem to affect anything but it kept repeating, and I couldn’t see any reason why. Since the idle hook updates the inlay hints anyway, I figured there wasn’t much to be gained by trying to debug it.

**Comment by shivjm** [comment]
I believe that addresses all the many issues in this small PR, ha. Should be ready to review again.

File: clients/lsp-javascript.el [hunk 3]
```diff
@@ -817,27 +822,48 @@ name (e.g. `data' variable passed as `data' parameter)."
...
-                                      (if whitespace-after? " " ""))))))))
-    :mode 'tick))
+         (let ((hints (lsp-get res :inlayHints)))
+           (when (> (length hints) 0)
+             (overlay-recenter
+              (-let* (([hint] hints)
+                      ((&javascript:InlayHint :position) hint))
+                (lsp--position-to-point position))))
+           (-each hints
+             #'(lambda (hint)
```
Comment (yyoncho) [inline clients/lsp-javascript.el:850]: this `#'` is no longer needed in modern elisp.

File: clients/lsp-javascript.el (hunk 1 above)
Comment (yyoncho) [reply clients/lsp-javascript.el:844]: looks ok to me, you may use `(unelss (seq-empty-p...`.

**Review by yyoncho — COMMENTED** [review]
Looks good to me. Some minor issues. 

we might consider implementing the lenses in a similar manner because we also have perf issues there.

**Review by jcs090218 — APPROVED** [review]

**Comment by shivjm** [comment]
Great, thanks for the review. I’ve pushed the latest changes. It should be good to go.

> we might consider implementing the lenses in a similar manner because we also have perf issues there.

Re: lenses, if it’s helpful, I could look at lsp-lens to see if I can figure out how to apply the same pattern.

**Comment by yyoncho** [comment]
Thank you!

**Comment by shivjm** [comment]
Thanks!

> That would be great. There are some differences(lenses are more complex because they can be lazy and there might be several lens providers) but I think that the code that places the lenses is compact.

Cool, I’ll take a look at it.

**Pull request merged**